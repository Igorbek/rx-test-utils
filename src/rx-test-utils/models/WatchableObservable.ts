import { Observable, Subject } from 'rxjs';
import { WatchNotification } from "./messages";

export class WatchableObservable<T> extends Observable<T> {
    subscriptions: Observable<Observable<WatchNotification<T>>> = new Subject<Observable<WatchNotification<T>>>();


}
