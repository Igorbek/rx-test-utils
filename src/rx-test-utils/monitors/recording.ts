import { TestMessage } from "rxjs/testing/TestMessage";
import { IScheduler } from 'rxjs/Scheduler';
import { Observable } from 'rxjs/Observable';
import { Notification } from 'rxjs/Notification';

import { IObservableMonitor, IObserverMonitor } from '../monitor';

interface RecordedMessageBase {
    time: number;
    order: number;
}

interface RecordedNextMessage<T> extends RecordedMessageBase {
    kind: 'N';
    value: T;
}

interface RecordedCompleteMessage extends RecordedMessageBase {
    kind: 'C';
}

interface RecordedErrorMessage extends RecordedMessageBase {
    kind: 'E';
    error: any;
}

type RecordedMessage<T> = RecordedNextMessage<T> | RecordedCompleteMessage | RecordedErrorMessage;

interface RecordedSubscription<T> {
    subscribedTime: number;
    subscribedOrder: number;
    unsubscribedTime?: number;
    unsubscribedOrder?: number;
    messages: RecordedMessage<T>[];
}

interface RecordedObservable<T> {
    subscriptions: RecordedSubscription<T>[];
}

export function createRecordingMonitor<T>(
    scheduler: IScheduler,
    timeOrderCoordinator: TimeOrderCoordinator): { monitor: IObservableMonitor<T>; recorded: RecordedObservable<T>; } {
    const recorded: RecordedObservable<T> = {
        subscriptions: []
    };

    const monitor: IObservableMonitor<T> = {
        subscribe() {
            const subscribedTime = scheduler.now(), subscribedOrder = timeOrderCoordinator.nextOrder(subscribedTime);
            const subscription: RecordedSubscription<T> = {
                subscribedTime,
                subscribedOrder,
                messages: []
            };
            recorded.subscriptions.push(subscription);

            return {
                next(value) {
                    const time = scheduler.now(), order = timeOrderCoordinator.nextOrder(time);
                    subscription.messages.push({
                        time,
                        order,
                        kind: 'N',
                        value
                    });
                },
                complete() {
                    const time = scheduler.now(), order = timeOrderCoordinator.nextOrder(time);
                    subscription.messages.push({
                        time,
                        order,
                        kind: 'C'
                    });
                },
                error(error) {
                    const time = scheduler.now(), order = timeOrderCoordinator.nextOrder(time);
                    subscription.messages.push({
                        time,
                        order,
                        kind: 'E',
                        error
                    });
                },
                unsubscribe() {
                    subscription.unsubscribedTime = scheduler.now();
                    subscription.unsubscribedOrder = timeOrderCoordinator.nextOrder(subscription.unsubscribedTime);
                }
            }
        }
    };

    return { monitor, recorded };
}

export interface TimeOrderCoordinator {
    nextOrder(time: number): number;
}

export function createTimeOrderCoordinator(): TimeOrderCoordinator {
    let lastTime = 0;
    let nextOrder = 0;
    return {
        nextOrder(time) {
            if (time < lastTime)
                throw new Error('Previous time was higher than passed.');
            if (time > lastTime) {
                lastTime = time;
                nextOrder = 0;
            }
            return nextOrder++;
        }
    }
}
