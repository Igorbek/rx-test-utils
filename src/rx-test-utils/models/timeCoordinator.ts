import { Scheduler } from 'rxjs/Scheduler'
import { TimeOrderCoordinator, TimeOrderAccessor } from './recording'

export function createTimeOrderCoordinator(): TimeOrderCoordinator {
    let lastTime = 0;
    let nextOrder = 0;
    return {
        nextOrder(time) {
            if (time < lastTime)
                throw new Error('Previous time was higher than passed.');
            if (time > lastTime) {
                lastTime = time;
                nextOrder = 0;
            }
            return nextOrder++;
        }
    }
}

export function createTimeOrderAccessor(
    scheduler: Scheduler,
    timeOrderCoordinator: TimeOrderCoordinator = createTimeOrderCoordinator()): TimeOrderAccessor {
    return {
        now() {
            const time = scheduler.now(), order = timeOrderCoordinator.nextOrder(time);
            return [time, order];
        }
    };
}
