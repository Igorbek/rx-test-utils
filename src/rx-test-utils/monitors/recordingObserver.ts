import { Observer } from 'rxjs/Observer'
import { Observable } from 'rxjs/Observable'
import { Scheduler } from 'rxjs/Scheduler';

import { WatchNotification } from '../models/messages'
import { RecordedObservable, RecordedSubscription, TimeOrderAccessor } from '../models/recording'

export function createRecordingObserver<T>(
    timeOrderAccessor: TimeOrderAccessor
): {
    observer: Observer<Observable<WatchNotification<T>>>;
    recorded: RecordedObservable<T>;
} {
    const recorded: RecordedObservable<T> = {
        subscriptions: []
    };

    const observer: Observer<Observable<WatchNotification<T>>> = {
        next(value) {
            const {
                observer: subscriptionObserver,
                recorded: recordedSubscription
            } = createRecordingSubscriptionObserver<T>(timeOrderAccessor);
            recorded.subscriptions.push(recordedSubscription);
            value.subscribe(subscriptionObserver);
        },
        complete() {
        },
        error(error) {
            throw error;
        }
    };

    return { observer, recorded };
}

export function createRecordingSubscriptionObserver<T>(
    timeOrderAccessor: TimeOrderAccessor
): {
    observer: Observer<WatchNotification<T>>;
    recorded: RecordedSubscription<T>;
} {
    const recorded: RecordedSubscription<T> = {
        notifications: [],
        subscribed: timeOrderAccessor.now()
    };

    const observer: Observer<WatchNotification<T>> = {
        next(notification) {
            const time = timeOrderAccessor.now()
            recorded.notifications.push({
                time,
                notification
            });
            if (notification.kind === 'U') {
                recorded.unsubscribed = time;
            }
        },
        complete() {
        },
        error(error) {
            throw error;
        }
    };

    return { observer, recorded };
}
