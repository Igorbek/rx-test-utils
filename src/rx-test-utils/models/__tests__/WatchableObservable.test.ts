import 'jest';

import { TestMessage } from "rxjs/testing/TestMessage";
import { IScheduler } from 'rxjs/Scheduler';
import { Observer, Observable } from 'rxjs'
import { TestScheduler } from 'rxjs/testing/TestScheduler'

import 'rxjs/add/operator/take';
import 'rxjs/add/operator/skip';

import { createTimeOrderAccessor } from '../timeCoordinator';
import { createRecordingObserver } from '../../monitors/recordingObserver';
import { toWatchable } from '../WatchableObservable'

describe('WatchableObservable', () => {
    it('it captures data', () => {
        const testScheduler = new TestScheduler((a, e) => a === e);
        const { observer, recorded } = createRecordingObserver(createTimeOrderAccessor(testScheduler));

        const source = testScheduler.createHotObservable('--1-2-3--4--5-')
        //const observable = monitored(Rx.Observable.interval(10, testScheduler).take(3), monitor);
        const observable = toWatchable(source.take(4));
        observable.subscriptions.subscribe(observer);

        const subscription = observable.subscribe();
        const subscription2 = observable.skip(1).subscribe();
        
        testScheduler.flush();

        expect(recorded).toMatchSnapshot();
    })
})
