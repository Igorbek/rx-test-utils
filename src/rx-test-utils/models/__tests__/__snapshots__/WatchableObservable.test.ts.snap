// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WatchableObservable it captures data 1`] = `
Object {
  "subscriptions": Array [
    Object {
      "notifications": Array [
        Object {
          "notification": Object {
            "kind": "N",
            "value": "1",
          },
          "time": Array [
            20,
            0,
          ],
        },
        Object {
          "notification": Object {
            "kind": "N",
            "value": "2",
          },
          "time": Array [
            40,
            0,
          ],
        },
        Object {
          "notification": Object {
            "kind": "N",
            "value": "3",
          },
          "time": Array [
            60,
            0,
          ],
        },
        Object {
          "notification": Object {
            "kind": "N",
            "value": "4",
          },
          "time": Array [
            90,
            0,
          ],
        },
        Object {
          "notification": Object {
            "kind": "C",
          },
          "time": Array [
            90,
            1,
          ],
        },
        Object {
          "notification": Object {
            "kind": "U",
          },
          "time": Array [
            90,
            2,
          ],
        },
      ],
      "subscribed": Array [
        0,
        0,
      ],
      "unsubscribed": Array [
        90,
        2,
      ],
    },
    Object {
      "notifications": Array [
        Object {
          "notification": Object {
            "kind": "N",
            "value": "1",
          },
          "time": Array [
            20,
            1,
          ],
        },
        Object {
          "notification": Object {
            "kind": "N",
            "value": "2",
          },
          "time": Array [
            40,
            1,
          ],
        },
        Object {
          "notification": Object {
            "kind": "N",
            "value": "3",
          },
          "time": Array [
            60,
            1,
          ],
        },
        Object {
          "notification": Object {
            "kind": "N",
            "value": "4",
          },
          "time": Array [
            90,
            3,
          ],
        },
        Object {
          "notification": Object {
            "kind": "C",
          },
          "time": Array [
            90,
            4,
          ],
        },
        Object {
          "notification": Object {
            "kind": "U",
          },
          "time": Array [
            90,
            5,
          ],
        },
      ],
      "subscribed": Array [
        0,
        1,
      ],
      "unsubscribed": Array [
        90,
        5,
      ],
    },
  ],
}
`;
